In [41]:
#Importing important libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [4]:
#Reading the data:
data= pd.read_csv('employee.csv')

In [5]:
#Reading the first 5 rows:
data.head()

name   age   income gender  department grade  performance_score
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723
1        S Kumar   NaN  16000.0      F     Finance    G0                520
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674
3      Ying Chin  45.0  65000.0      F       Sales    G3                556
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711

There are missing data in the age column. Some data are numeric (income, age, and performance score) and some are categoric (name, grade and gender).

In [6]:
#Looking at the last 5 rows:
data.tail()

name   age   income gender  department grade  performance_score
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711
5  Satyam Sharma   NaN  62000.0    NaN       Sales    G3                649
6   James Authur  54.0      NaN      F  Operations    G3                 53
7     Josh Wills  54.0  52000.0      F     Finance    G3                901
8       Leo Duck  23.0  98000.0      M       Sales    G4                709

In [7]:
#Printing list of columns and can be useful when dropping some columns for feature analysis:
data.columns

Index(['name', 'age', 'income', 'gender', 'department', 'grade',
       'performance_score'],
      dtype='object')

In [8]:
#Print the number of data rows and columns
data.shape

(9, 7)

The data has 9 rows and 7 columns.

In [9]:
#Checking the information in the dataframe:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   name               9 non-null      object 
 1   age                7 non-null      float64
 2   income             7 non-null      float64
 3   gender             7 non-null      object 
 4   department         9 non-null      object 
 5   grade              9 non-null      object 
 6   performance_score  9 non-null      int64  
dtypes: float64(2), int64(1), object(4)
memory usage: 636.0+ bytes


Some data type are objects, some numeric (int64 and float64). There is missing data in age, income, and gender because they have 7 out 9 non null values.

In [10]:
#Descriptive statistics:
data.describe()

age        income  performance_score
count   7.000000      7.000000           9.000000
mean   40.428571  52857.142857         610.666667
std    12.204605  26028.372797         235.671912
min    23.000000  16000.000000          53.000000
25%    31.000000  38500.000000         556.000000
50%    45.000000  52000.000000         674.000000
75%    49.500000  63500.000000         711.000000
max    54.000000  98000.000000         901.000000

The minimum age of employs ranges from 23 to 54 years. The mean age is 40.4 years and the majority of the people have ages spread around 40.4-12.2 (30.2- 52.6 years) because the standard deviation for the age is 12.2 years. &5% of the employees earn less than £63 000. 

In [11]:
#Data filtering and column-wise:
data.filter(['name' , 'department'])

name  department
0    Allen Smith  Operations
1        S Kumar     Finance
2    Jack Morgan     Finance
3      Ying Chin       Sales
4  Dheeraj Patel  Operations
5  Satyam Sharma       Sales
6   James Authur  Operations
7     Josh Wills     Finance
8       Leo Duck       Sales

This would return only the columns with the name of the employee and thier department.

In [13]:
data.filter(['name', 'performance_score'])

name  performance_score
0    Allen Smith                723
1        S Kumar                520
2    Jack Morgan                674
3      Ying Chin                556
4  Dheeraj Patel                711
5  Satyam Sharma                649
6   James Authur                 53
7     Josh Wills                901
8       Leo Duck                709

James Arthur has the lowest perfomance score of 53.

In [14]:
#This selects a single  column and makes a pandas series list.
data.filter(['name'])

name
0    Allen Smith
1        S Kumar
2    Jack Morgan
3      Ying Chin
4  Dheeraj Patel
5  Satyam Sharma
6   James Authur
7     Josh Wills
8       Leo Duck

In [18]:
#This selects a single column and makes a python list or pandas dataframe:
data[['name']]

name
0    Allen Smith
1        S Kumar
2    Jack Morgan
3      Ying Chin
4  Dheeraj Patel
5  Satyam Sharma
6   James Authur
7     Josh Wills
8       Leo Duck

In [19]:
#Filtering two columns to output a dataframe:
data[['name', 'department']]

name  department
0    Allen Smith  Operations
1        S Kumar     Finance
2    Jack Morgan     Finance
3      Ying Chin       Sales
4  Dheeraj Patel  Operations
5  Satyam Sharma       Sales
6   James Authur  Operations
7     Josh Wills     Finance
8       Leo Duck       Sales

In [25]:
#Filtering row-wise:
#Selecting rows for chosen indices:
data.filter([0, 1,2], axis=0)

name   age   income gender  department grade  performance_score
0  Allen Smith  45.0      NaN    NaN  Operations    G3                723
1      S Kumar   NaN  16000.0      F     Finance    G0                520
2  Jack Morgan  32.0  35000.0      M     Finance    G2                674

In [26]:
#Filter data by slicing:
data[2:5]

name   age   income gender  department grade  performance_score
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674
3      Ying Chin  45.0  65000.0      F       Sales    G3                556
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711

In [28]:
data['department']

0    Operations
1       Finance
2       Finance
3         Sales
4    Operations
5         Sales
6    Operations
7       Finance
8         Sales
Name: department, dtype: object

In [30]:
data[data.department=='Sales']

name   age   income gender department grade  performance_score
3      Ying Chin  45.0  65000.0      F      Sales    G3                556
5  Satyam Sharma   NaN  62000.0    NaN      Sales    G3                649
8       Leo Duck  23.0  98000.0      M      Sales    G4                709

In [33]:
#Filtering multiple columns:
data[data.department.isin(['Sales', 'Finance'])]

name   age   income gender department grade  performance_score
1        S Kumar   NaN  16000.0      F    Finance    G0                520
2    Jack Morgan  32.0  35000.0      M    Finance    G2                674
3      Ying Chin  45.0  65000.0      F      Sales    G3                556
5  Satyam Sharma   NaN  62000.0    NaN      Sales    G3                649
7     Josh Wills  54.0  52000.0      F    Finance    G3                901
8       Leo Duck  23.0  98000.0      M      Sales    G4                709

This line of code is filtering a DataFrame named data based on the values in the department column. Here's a breakdown:

data: This refers to the DataFrame object that contains the data.
data.department: This refers to the department column in the DataFrame data.
data.department.isin(['Sales', 'Finance']): This is a boolean expression that checks whether each value in the department column is either 'Sales' or 'Finance'. The isin() method checks if each value in the column is contained within the provided list ['Sales', 'Finance'].
data[data.department.isin(['Sales', 'Finance'])]: This uses boolean indexing to filter rows from the DataFrame data where the condition data.department.isin(['Sales', 'Finance']) is True. So, it selects only those rows where the department is either 'Sales' or 'Finance'.

In [34]:
#Filtering employees who have performance scores greater or equal to 700:
data[data.performance_score>=700]

name   age   income gender  department grade  performance_score
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711
7     Josh Wills  54.0  52000.0      F     Finance    G3                901
8       Leo Duck  23.0  98000.0      M       Sales    G4                709

In [37]:
#Filtering employees who have performance scores greater or equal to 500 but less than 700:
data[(data.performance_score >= 500) & (data.performance_score < 700)]

name   age   income gender department grade  performance_score
1        S Kumar   NaN  16000.0      F    Finance    G0                520
2    Jack Morgan  32.0  35000.0      M    Finance    G2                674
3      Ying Chin  45.0  65000.0      F      Sales    G3                556
5  Satyam Sharma   NaN  62000.0    NaN      Sales    G3                649

In [38]:
#Query performance of emeployees who have scores less than 500:
data.query('performance_score < 500')

name   age  income gender  department grade  performance_score
6  James Authur  54.0     NaN      F  Operations    G3                 53

Dealing with missing values

In [46]:
#Drop all null values using dropna():
data.dropna()

name   age   income gender  department grade  performance_score
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674
3      Ying Chin  45.0  65000.0      F       Sales    G3                556
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711
7     Josh Wills  54.0  52000.0      F     Finance    G3                901
8       Leo Duck  23.0  98000.0      M       Sales    G4                709

In [47]:
data

name   age   income gender  department grade  performance_score
0    Allen Smith  45.0      NaN    NaN  Operations    G3                723
1        S Kumar   NaN  16000.0      F     Finance    G0                520
2    Jack Morgan  32.0  35000.0      M     Finance    G2                674
3      Ying Chin  45.0  65000.0      F       Sales    G3                556
4  Dheeraj Patel  30.0  42000.0      F  Operations    G2                711
5  Satyam Sharma   NaN  62000.0    NaN       Sales    G3                649
6   James Authur  54.0      NaN      F  Operations    G3                 53
7     Josh Wills  54.0  52000.0      F     Finance    G3                901
8       Leo Duck  23.0  98000.0      M       Sales    G4                709

In [49]:
#Fill in any missing values in age column with mean values of the age:
data['age']=data.age.fillna(data.age.mean())

In [50]:
data

name        age   income gender  department grade  \
0    Allen Smith  45.000000      NaN    NaN  Operations    G3   
1        S Kumar  40.428571  16000.0      F     Finance    G0   
2    Jack Morgan  32.000000  35000.0      M     Finance    G2   
3      Ying Chin  45.000000  65000.0      F       Sales    G3   
4  Dheeraj Patel  30.000000  42000.0      F  Operations    G2   
5  Satyam Sharma  40.428571  62000.0    NaN       Sales    G3   
6   James Authur  54.000000      NaN      F  Operations    G3   
7     Josh Wills  54.000000  52000.0      F     Finance    G3   
8       Leo Duck  23.000000  98000.0      M       Sales    G4   

   performance_score  
0                723  
1                520  
2                674  
3                556  
4                711  
5                649  
6                 53  
7                901  
8                709

In [53]:
#Fill na values with median:
new_data=data['income']=data.income.fillna(data.income.median())

In [54]:
new_data

0    52000.0
1    16000.0
2    35000.0
3    65000.0
4    42000.0
5    62000.0
6    52000.0
7    52000.0
8    98000.0
Name: income, dtype: float64

In [55]:
data

name        age   income gender  department grade  \
0    Allen Smith  45.000000  52000.0    NaN  Operations    G3   
1        S Kumar  40.428571  16000.0      F     Finance    G0   
2    Jack Morgan  32.000000  35000.0      M     Finance    G2   
3      Ying Chin  45.000000  65000.0      F       Sales    G3   
4  Dheeraj Patel  30.000000  42000.0      F  Operations    G2   
5  Satyam Sharma  40.428571  62000.0    NaN       Sales    G3   
6   James Authur  54.000000  52000.0      F  Operations    G3   
7     Josh Wills  54.000000  52000.0      F     Finance    G3   
8       Leo Duck  23.000000  98000.0      M       Sales    G4   

   performance_score  
0                723  
1                520  
2                674  
3                556  
4                711  
5                649  
6                 53  
7                901  
8                709

In [64]:
#Fill in all the missing values in gender column with the mode for the gender values:
data['gender']=data['gender'].fillna(data['gender'].mode()[0])

In [65]:
data

name        age   income gender  department grade  \
0    Allen Smith  45.000000  52000.0      F  Operations    G3   
1        S Kumar  40.428571  16000.0      F     Finance    G0   
2    Jack Morgan  32.000000  35000.0      M     Finance    G2   
3      Ying Chin  45.000000  65000.0      F       Sales    G3   
4  Dheeraj Patel  30.000000  42000.0      F  Operations    G2   
5  Satyam Sharma  40.428571  62000.0      F       Sales    G3   
6   James Authur  54.000000  52000.0      F  Operations    G3   
7     Josh Wills  54.000000  52000.0      F     Finance    G3   
8       Leo Duck  23.000000  98000.0      M       Sales    G4   

   performance_score  
0                723  
1                520  
2                674  
3                556  
4                711  
5                649  
6                 53  
7                901  
8                709